geoutils/spatial_tools.py

```diff
@@ -88,6 +88,69 @@ def get_valid_extent(array: np.ndarray | np.ma.masked_array) -> tuple[int, ...]:
     return rows_nonzero[0], rows_nonzero[-1], cols_nonzero[0], cols_nonzero[-1]
 
 
+def load_multiple_rasters(raster_paths: list[str], ref_grid: int | None = None, **kwargs: Any) -> list[RasterType]:
+    """
+    Function to load multiple rasters at once in a memory efficient way.
+    First the area of overlap is estimated from metadata only.
+    The rasters are cropped to this area of overlap and loaded into memory.
+    Optionally, the rasters are all reprojected to the grid of one raster set as reference.
+
+    :param raster_paths: List of paths to the rasters to be loaded
+    :param ref_grid: If set to an integer value, the raster with that index will be considered as the reference
+    and all other rasters will be reprojected on the same grid
+    :param kwargs: optional arguments to be passed to Raster.reproject, e.g. the resampling method
+
+    :returns: a list of loaded Raster instances
+    """
+    # Check ref_grid value. If None, use first raster for reference CRS, but does not reproject
+    if isinstance(ref_grid, int):
+        reproject = True
+    elif ref_grid is None:
+        ref_grid = 0
+        reproject = False
+    else:
+        raise ValueError("`ref_grid` must be None or an integer")
+
+    # Need to define a reference CRS for calculating intersection
+    ref_crs = gu.Raster(raster_paths[ref_grid], load_data=False).crs
+
+    # First load all rasters metadata
+    output_rst = []
+    bounds = []
+    for path in raster_paths:
+        # Initialize raster
+        rst = gu.Raster(path, load_data=False)
+        output_rst.append(rst)
+
+        # Get bound in reference CRS
+        bound = rst.get_bounds_projected(ref_crs)
+        bounds.append(bound)
+
+    # Second get the intersection of all raster bounds and crop rasters accordingly
+    intersection = gu.projtools.merge_bounds(bounds, "intersection")
+    if intersection == ():
+        warnings.warn("No intersection found, returning unloaded rasters")
+        return output_rst
+
+    for rst in output_rst:
+        # Calculate bounds in rst's CRS
+        new_bounds = rio.warp.transform_bounds(
+            ref_crs, rst.crs, intersection[0], intersection[1], intersection[2], intersection[3], densify_pts=5000
+        )
+        # Ensure bounds align with the original ones, to avoid resampling at this stage
+        new_bounds = gu.projtools.align_bounds(rst.transform, new_bounds)
+        rst.crop(new_bounds, mode="match_pixel")
```

Comment:
So crop is memory efficient, doesn't load all the entire Raster?

Reply:
If the cropping exactly matches the extent, won't it negatively affect the reprojection later? Like, the 1/2 pixels on the edge might end up as `nodata` because of the resampling, while they wouldn't have been if using the entire original Raster?

Reply:
Yes, crop only loads the data within the corresponding window. That's why it's useful to crop the raster, especially if one is much larger.
Actually, reproject does not need to have the data loaded anymore, so if followed by reproject, one does not need to crop and load the data. I will try that instead. If the data is not loaded, it uses the raster's band instead, so rasterio handles all this for us.

Reply:
Solved in [df1bad1](https://github.com/GlacioHack/geoutils/pull/307/commits/df1bad1df509dee2f8f2cb4f31a7cf2af34d157b)